### PR TITLE
Donat/problems/fileloc index

### PR DESCRIPTION
**Pull request by donat**

### CODE REVIEW

**Comment by donat** [comment]
@bot-gradle test RFM

**Comment by donat** [comment]
@bot-gradle test RFM

**Comment by donat** [comment]
@bot-gradle test RFM

**Comment by donat** [comment]
@bot-gradle test RFM

File: platforms/ide/problems-api/src/main/java/org/gradle/api/problems/ProblemSpec.java [hunk 1]
```diff
@@ -18,87 +18,137 @@
...
+     * <p>
+     * The label is the main, human-readable representation of the problem.
+     * It is a mandatory property to configure when emitting a problem with {@link ProblemReporter}.
      *
-     * @return the builder for the next required property
+     * @param label the short message
+     * @return this
      * @since 8.6
      */
-    ProblemSpec label(String label, Object... args);
```
Comment (hegyibalint) [inline platforms/ide/problems-api/src/main/java/org/gradle/api/problems/ProblemSpec.java:48]: 🤔 I think here we should use a typical `Logger`-like API:
 - A `label(String label)` method for messages not requiring any formatting
 - A `label(String label, Objects... args)` method for messages **requiring** formatting.

I think this is the best solution, which would solve both of our concerns about preventing an Exception from `String.format`, but at the same time, maximizing DX.

File: platforms/ide/tooling-api/src/main/java/org/gradle/tooling/events/problems/OffsetInFileLocation.java [hunk 2]
```diff
@@ -14,52 +14,33 @@
...
-    @Nullable
-    Integer getLine();
+public interface OffsetInFileLocation extends FileLocation {
 
     /**
-     * The offset on the selected line.
+     * The global offset from the beginning of the file.
      *
-     * @return the column
+     * @return the zero-indexed the offset
```
Comment (hegyibalint) [inline platforms/ide/tooling-api/src/main/java/org/gradle/tooling/events/problems/OffsetInFileLocation.java:34]:
```suggestion
     * @return the zero-indexed offset
```

**Review by hegyibalint — CHANGES_REQUESTED** [review]
Found some doc issue, otherwise LGTM!

File: platforms/ide/tooling-api/src/main/java/org/gradle/tooling/events/problems/ProblemDescriptor.java [hunk 3]
```diff
@@ -32,15 +32,6 @@
...
 public interface ProblemDescriptor extends OperationDescriptor {
 
-    /**
-     * Mock for the problem properties in JSON format.
-     * <p>
-     * Note: This method is not functional anymore and will not return anything useful.
-     *
-     * @return the problem properties.
-     */
-    String getJson(); // TODO https://github.com/gradle/gradle/issues/27125
```
Comment (alllex) [inline platforms/ide/tooling-api/src/main/java/org/gradle/tooling/events/problems/ProblemDescriptor.java:42]: 👍

File: platforms/ide/tooling-api/src/main/java/org/gradle/tooling/events/problems/LineInFileLocation.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+ *
+ * @since 8.6
+ */
+@Incubating
+public interface LineInFileLocation extends FileLocation {
+
+    /**
+     * The line number within the file.
+     * <p>
+     * The column is one-indexed, i.e. the first line in the file is line number 1.
```
Comment (alllex) [inline platforms/ide/tooling-api/src/main/java/org/gradle/tooling/events/problems/LineInFileLocation.java:34]:
```suggestion
     * The line number is one-indexed, i.e. the first line in the file is line number 1.
```

File: platforms/ide/problems-api/src/main/java/org/gradle/api/problems/ProblemSpec.java (hunk 1 above)
Comment (alllex) [reply platforms/ide/problems-api/src/main/java/org/gradle/api/problems/ProblemSpec.java:48]: I don't think that baking in `String.format` contract for a string is a good idea. The requirements for the format string are huge. I agree, it's less of a problem when it's just an overload, but still I don't think it's justified.

I would argue that the DX is actually better if the user decides to call `String.format` themselves. For one, IDEA even provides static analysis that makes sure your format string matches the number of arguments that is provided. By making the `format` call hidden inside `label` we take away these insights from the user.

Moreover, with time, there are more and more ways to interpolate a string. Kotlin has interpolation from the beginning, and it also provides functions like `buildString`. Even [Java introduces interpolation](https://openjdk.org/jeps/430), hinting that `String.format` is not always a perfect fit.

The loggers have such an API primarily for performance reasons, which I don't believe is applicable for Problems API. Also, loggers are [known to come up with their own formats](https://www.slf4j.org/faq.html#logging_performance), which dilutes the idea of universal expectation, and forces the users of the Problems API to read the javadocs (nobody reads javadoc ;)).

In other words, I think our API should not be biased towards `String.format`.

File: platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/LineInFileLocation.java [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+     * The column is one-indexed, i.e. the first line in the file is line number 1.
+     *
+     * @return the line number
+     */
+    int getLine();
+
+    /**
+     * The starting column on the selected line.
+     * <p>
+     * The column is one-indexed, i.e. the first line in the file is line number 1.
```
Comment (alllex) [inline platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/LineInFileLocation.java:39]:
```suggestion
     * The column is <b>one-indexed</b>, i.e. the first line in the file is line number 1.
```

**Review by alllex — APPROVED** [review]

File: platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/LineInFileLocation.java (hunk 5 above)
Comment (donat) [reply platforms/ide/problems-api/src/main/java/org/gradle/api/problems/internal/LineInFileLocation.java:39]: Fixed

File: platforms/ide/problems-api/src/main/java/org/gradle/api/problems/ProblemSpec.java (hunk 1 above)
Comment (hegyibalint) [reply platforms/ide/problems-api/src/main/java/org/gradle/api/problems/ProblemSpec.java:48]: I think at this point the logger format (especially after Java 8) is a bit of a moot point, and it is

Comment (hegyibalint) [reply platforms/ide/problems-api/src/main/java/org/gradle/api/problems/ProblemSpec.java:48]: I've mentioned to @donat that I won't die on this hill.

I think, at this point, the logger format (especially after Java 8) is a convention rather than a limitation. It is indeed true that packing `String.format(...)`'s semantics is a heavy one; though as being used often by the developers, it's a well-known one. 

> For one, IDEA even provides static analysis that makes sure your format string matches the number of arguments that is provided.

True!

> Moreover, with time, there are more and more ways to interpolate a string. Kotlin has interpolation from the beginning, and it also provides functions like buildString. Even [Java introduces interpolation](https://openjdk.org/jeps/430), hinting that String.format is not always a perfect fit.

This is a bit optimistic view: how many of our plugin developers use Kotlin, or JDK21 (though the LTS situation might be helpful to tip avaw from JDK11, I'm not holding my breath)

---

In general, I agree with the points presented here, and let's go with the template-less solution!

Comment (wolfs) [reply platforms/ide/problems-api/src/main/java/org/gradle/api/problems/ProblemSpec.java:48]: I agree that it is much simpler API wise to leave that up to the users of the API how they format their string. The only reason to add such an API would be performance, since it might be quite expensive to create a lot of dynamic strings. On the other hand,  I am not sure we could avoid the creation of the `String`s just because we offer such an API.

**Comment by donat** [comment]
@bot-gradle test this

**Comment by donat** [comment]
@bot-gradle test and merge
